### DISCUSSION TITLE
Welcome to Sharp-Grad Discussions!

### DISCUSSION THREAD

**Discussion by Leonardo16AM** [Leonardo16AM/Sharp-Grad#8] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by DevNullx64** [comment]
Hi :)
From my side I've some check point. Use vectorisation, multithreading, and GPU. I whish run RetNet (or just Transformers) on my hown personal computer for less ressources cost and hover heat as possible.
Most should be done in C++ instead of C#. If It's needed I will do a litle but I know why this beautiful language must be avoid as possible.

Cheer

**Reply by Leonardo16AM** [reply]
There is an specific transformer architecture that you want to run? I suppose that you want to train it, right? I can help with the C++ part, if you feel comfortable we can migrate the Project to C++ if you wanted to. 
What are your expectations about the project or what are your final goals? There is an specific reason why you prefer it to be in C#, instead of using Python libraries?  I ask just for curiosity, if you prefer we can stay using C#.

**Reply by DevNullx64** [reply]
I've no specific transformer architecture in mind. LLM or Image Generator. But yes ofcourse, that can be train :)

C++ is a low-level language. While it allows for great speed, it also comes with inherent security risks by design. Low-level access requires careful management. C# uses a garbage collector (GC) to handle memory leaks and includes buffer overflow mitigations. C# is designed to be "safe." Many leaks in C++ are due to poor assertions. For example, it is easier to write a malicious Virtual Memory Page in C++ to unlock the kernel than in C#.

C# is a "safe" language. It manages memory allocation and mitigates some buffer overflow attacks. It is not compiled into a native machine language instruction set directly. Instead, .NET uses the Common Intermediate Language (CIL) as a virtual machine assembler. The compiler performs syntax analysis, dereferencing, and immediate optimization to generate CIL. Translating CIL to machine language is very fast with minimal overhead, resulting in native machine language executable code.

Python is the best language in the world for programming, but it is not always suitable for large-scale development. It's like comparing an Excel/VBA file to an SQL server. Python uses an interpreter that reads a text file and performs the necessary compilation steps on-the-fly. While you can cache the results to speed up execution, this is a workaround. Python is extremely permissive and flexible, placing the responsibility for code quality on the developer. For small projects, Python is highly recommended initially. However, as the project grows, the cognitive load on the developer increases rapidly. In typed languages, the responsibility for code quality is partially transferred to the type descriptions within the code, making it more complex but also more secure.

My expectation is to have a real machine learning framework in safe environmenet whit lesss overheat as possible. But why not plan a Python interface to be interoperable ?

**Reply by DevNullx64** [reply]
And for C++ part, I need to check if C++/CLI is really dead or just not on the behind of the scene

edit:
not it's good. So If we need C++, I recomande to use C++/CLI instead

**Reply by Leonardo16AM** [reply]
Ok, I think that C# is ok, we don't need to complicate our lives with C++ right now.

**Reply by DevNullx64** [reply]
Right :)

I esitate to use C++ because I try to hack the GC memory that handle Array bound. Currently I've recode with Shape something that the CPU have as hardvware and C# insoftware already (for outofbound exception by exemple). The BOUND instruction. For sure C# in native Array use this CPU instruction on use profit for some optimisation in the computation of the index. the x86 architecture have a lot of way to compute adresses. So, rewritre the memory of the GC for an Array can lead to an Array rechape as zero memory copy. If I found the good way, I will implemente this.
